File: crates/idakit-sys-codegen/src/domains-tmp/range.rs
use super::super::model::*;
use super::EA;

/// The function-range domain: the SDK POD `range_t` bound as a `Trivial` `ExternType` that crosses
/// by value four ways (bare return, by-value argument, shared-struct field, and `Vec` element). All
/// bodies are hand-written in `facade/range_custom.cc` (they iterate a `func_tail_iterator_t`).
pub const RANGE: Domain = Domain {
    name: "range",
    sdk_includes: &["<funcs.hpp>", "<range.hpp>", "<stdexcept>"],
    externs: &[ExternTy {
        rust_name: "RangeT",
        cxx_name: "range_t",
        kind: ExternKind::Trivial(fields! {
            start: U64 = "`start_ea`, inclusive.";
            end: U64 = "`end_ea`, exclusive.";
        }),
        doc: "A `#[repr(C)]` mirror of the SDK's `range_t`, crossing the bridge by value as a \
              `Trivial` `ExternType`.",
        safety: "RangeT's two u64 fields mirror range_t's two ea_t members under __EA64__, and \
                 range_t is trivially move-constructible and destructible, so it crosses by value \
                 soundly. cxx re-checks the triviality half with a C++ static_assert.",
    }],
    structs: &[SharedStruct {
        name: "ChunkInfo",
        doc: "One function chunk: its index paired with its address range.",
        fields: fields! {
            index: Usize = "Zero-based chunk index (the entry chunk is `0`).";
            range: Extern("RangeT") = "The chunk's address range.";
        },
    }],
    custom_tu: Some("facade/range_custom.cc"),
    body_helpers: None,
    fns: &[
        FnSpec {
            name: "range_entry_chunk",
            receiver: None,
            args: EA,
            ret: RetKind::ResultExtern("RangeT"),
            body: BodyKind::Custom,
            doc: "Entry chunk (index `0`) of the function containing `ea`, returned by value; \
                  `Err` when no function is there.",
        },
        FnSpec {
            name: "range_size",
            receiver: None,
            args: args!(r: Extern("RangeT")),
            ret: RetKind::U64,
            body: BodyKind::Custom,
            doc: "Size (`end - start`) of a `range_t` passed by value.",
        },
        FnSpec {
            name: "range_chunk_info",
            receiver: None,
            args: args!(ea: U64, n: Usize),
            ret: RetKind::ResultShared("ChunkInfo"),
            body: BodyKind::Custom,
            doc: "Chunk `n` of the function at `ea` as a `ChunkInfo`; `Err` when `n` is out of \
                  range.",
        },
        FnSpec {
            name: "range_all_chunks",
            receiver: None,
            args: EA,
            ret: RetKind::ResultVec("RangeT"),
            body: BodyKind::Custom,
            doc: "Every chunk (entry plus tails) of the function at `ea` as one owned `Vec`; \
                  `Err` when no function is there.",
        },
    ],
};

File: crates/idakit-sys/facade/range.cpp
// Hand-written Custom bodies for the generated range domain (namespace gen). range_t is a
// Trivial ExternType, so it crosses the cxx bridge by value in every direction: returned bare,
// taken by value, embedded in the ChunkInfo shared struct, and collected into an owned
// rust::Vec<range_t>. A missing function or an out-of-range chunk index throws, surfacing as a
// Rust Err.

#include <ida.hpp>
#include <pro.h>

#include <funcs.hpp> // function_tail_iterator_t
#include <range.hpp>

#include <stdexcept>

#include "gen_range.h"
// The cxx-generated header defines the ChunkInfo shared struct (full definition needed to construct
// it) and instantiates rust::Vec<range_t>; gen_range.h only forward-declares ChunkInfo.
#include "gen_bridge.h"

#include "compat.h"

namespace gen {

// The entry (main) chunk of the function at addr; throws if there's no function there or it has
// no entry chunk.
::range_t range_entry_chunk(uint64_t addr) {
  function_tail_iterator_t fti;
  if (!fti.set(static_cast<ea_t>(addr)))
    throw std::out_of_range("no function at address");
  if (!fti.main())
    throw std::out_of_range("function has no entry chunk");
  ::range_t chunk;
  fti.chunk(&chunk);
  return chunk;
}

// The byte length of range (end_ea - start_ea).
uint64_t range_size(::range_t range) { return static_cast<uint64_t>(range.size()); }

// The n-th chunk (entry chunk first, then tails) of the function at addr, paired with its index
// in ChunkInfo; throws if there's no function at addr or n is out of range.
ChunkInfo range_chunk_info(uint64_t addr, size_t n) {
  function_tail_iterator_t fti;
  if (!fti.set(static_cast<ea_t>(addr)))
    throw std::out_of_range("no function at address");
  size_t i = 0;
  for (bool ok = fti.main(); ok; ok = fti.next(), i++) {
    if (i == n) {
      ChunkInfo info;
      info.index = n;
      fti.chunk(&info.range);
      return info;
    }
  }
  throw std::out_of_range("chunk index out of range");
}

// Every chunk of the function at addr (entry chunk first, then tails), collected into an owned
// Vec; throws if there's no function at addr.
rust::Vec<::range_t> range_all_chunks(uint64_t addr) {
  function_tail_iterator_t fti;
  if (!fti.set(static_cast<ea_t>(addr)))
    throw std::out_of_range("no function at address");
  rust::Vec<::range_t> out;
  ::range_t chunk;
  for (bool ok = fti.main(); ok; ok = fti.next()) {
    fti.chunk(&chunk);
    out.push_back(chunk);
  }
  return out;
}

} // namespace gen
